cloudsync/smartsync.py

```diff
@@ -0,0 +1,384 @@
+        time.sleep(.01)  # give smartsync a chance to preempt
+        with self._mutex:
+            super().do()
+
+    def register_auto_sync_callback(self, callback: Callable):
+        self.state.register_auto_sync_callback(callback)
+
+    def _ent_to_smartinfo(self, ent: Optional[SyncEntry], local_dir_info: Optional[DirInfo], local_path) -> SmartInfo:  # pylint: disable=too-many-locals
+        assert ent or local_dir_info, "must provide one of ent or local_dir_info"
+
+        if local_dir_info:  # file is synced, use the local info
+            otype = local_dir_info.otype
+            oid = local_dir_info.oid
+            remote_oid = ent[REMOTE].oid if ent else None
+            obj_hash = local_dir_info.hash
+            path = local_dir_info.path or ent[LOCAL].path
+            size = local_dir_info.size
+            name = local_dir_info.name
+            mtime = local_dir_info.mtime
+            is_synced = True
+        else:
+            otype = ent[REMOTE].otype
+            oid = ent[LOCAL].oid
+            remote_oid = ent[REMOTE].oid
+            obj_hash = None
+            path = local_path
+            size = ent[REMOTE].size
+            name = self.providers[REMOTE].basename(ent[REMOTE].path)
+            mtime = ent[REMOTE].mtime
+            is_synced = False
+        shared = False  # TODO: ent[REMOTE].shared,
+        readonly = False  # TODO: ent[REMOTE].readonly
+        retval = SmartInfo(otype=otype,
+                           oid=oid,
+                           remote_oid=remote_oid,
+                           hash=obj_hash,
+                           path=path,
+                           size=size,
+                           name=name,
+                           mtime=mtime,
+                           shared=shared,  # TODO: ent[REMOTE].shared,
+                           readonly=readonly,  # TODO: ent[REMOTE].readonly
+                           is_synced=is_synced,
+                           )
+        return retval
+
+    def _sync_one_entry(self, sync: SyncEntry):
+        try:
+            something_got_done = self.smgr.pre_sync(sync)
+            if not something_got_done:
+                something_got_done = self.smgr.sync(sync, want_raise=True)
+            self.state.storage_commit()
+            return something_got_done
+        except ex.CloudException as e:
+            path = sync[REMOTE].path if sync[REMOTE].path else sync[LOCAL].path
+            self.nmgr.notify_from_exception(SourceEnum.SYNC, e, path)
+            raise
+
+    def _smart_unsync_ent(self, ent):
+        if ent:
+            self.state.unconditionally_get_latest(ent, LOCAL)
+            if ent[LOCAL].hash != ent[LOCAL].sync_hash or ent[LOCAL].parent.paths_differ(LOCAL):
+                ent[LOCAL].changed = ent[LOCAL].changed or time.time()
+                self._sync_one_entry(ent)
+
+            found = self.state.smart_unsync_ent(ent)
+            if found:
+                self._smart_unsync_ent(found)
```

Comment:
Does this need to be recursive?

Reply:
Good catch -- this logic is not correct, I'll fix

Reply:
It turns out the behavior of this version of the code is coincidentally the intended behavior, albeit taking a rather circuitous route through the code. I've fixed it by removing the recursion.